In [3]:
import pandas as pd

# I.DF 불러오기

In [4]:
market_11  = pd.read_excel("../dataset_2nd/market_11.xlsx")
market_12  = pd.read_excel("../dataset_2nd/market_12.xlsx")
market_13  = pd.read_excel("../dataset_2nd/market_13.xlsx")
market_14  = pd.read_excel("../dataset_2nd/market_14.xlsx")
market_15  = pd.read_excel("../dataset_2nd/market_15.xlsx")
market_16  = pd.read_excel("../dataset_2nd/market_16.xlsx")
market_17  = pd.read_excel("../dataset_2nd/market_17.xlsx")
market_18  = pd.read_excel("../dataset_2nd/market_18.xlsx")
market_19  = pd.read_excel("../dataset_2nd/market_19.xlsx")
market_20  = pd.read_excel("../dataset_2nd/market_20.xlsx")
market_21  = pd.read_excel("../dataset_2nd/market_21.xlsx")
market_22  = pd.read_excel("../dataset_2nd/market_22.xlsx")

In [5]:
# z_score 때, 주식수 & 종가 -> 시가총액 만든 DataFrame ( 결측치 = 주식수:1875, 종가:3237, 시총:3478)
df_mk = pd.read_excel('../dataset_2nd/mkcap.xlsx', index_col=False)
df_mk.drop(df_mk.columns[0], axis =1, inplace=True )

## I-I. DF concat

In [7]:
df_price = pd.concat([market_11,market_12,market_13,market_14,market_15,market_16,market_17,market_18,market_19,market_20,market_21,market_22], axis=0)

In [9]:
# df_price : Jiwoo

# object
# price['거래소코드'].info()

# 3093
len(df_price['거래소코드'].unique())

# 1954
len(df_price['회사명'].unique())

1954

In [10]:
# df_code : ZiNo

# int
# df_code['거래소코드'].info()

# 1818
print(len(df_mk['거래소코드'].astype('str').unique()))

# 1795 
len(df_mk['회사명'].astype('str').unique())

1818


1795

In [11]:
# Check
# df_price.head()

---

# II. 시가총액 Market Capitalization
---
### 1. df_mk의 필요한 컬러만 따로 df : mk_real (Market Capitalization, 시가총액)
    - df_mk에서 구할 수 있는 시총은 일단 확보 



In [12]:
# 1. mk_real 데이터프레임 : 기확보 시가총액(결측치 존재) from df_mk
mk_real = df_mk.iloc[:,[0,1,2,3,-2,-1]]

In [13]:
mk_real

회사명  거래소코드     회계년도  [제조]* 발행한 주식총수(*)(IFRS연결)(주)   종가(원)  \
0      (주)CMG제약  58820  2014/12                    86626866.0   912.0   
1      (주)CMG제약  58820  2018/12                   108887244.0  4235.0   
2      (주)CMG제약  58820  2019/12                   138892244.0  4250.0   
3      (주)CMG제약  58820  2020/12                   138892244.0  5780.0   
4      (주)CMG제약  58820  2017/12                   108887244.0  6240.0   
...         ...    ...      ...                           ...     ...   
17482   흥구석유(주)  24060  2022/12                    15000000.0  5680.0   
17483   흥구석유(주)  24060  2013/12                    15000000.0  2000.0   
17484   흥구석유(주)  24060  2018/12                    15000000.0  2705.0   
17485   흥구석유(주)  24060  2016/12                    15000000.0  3425.0   
17486   흥구석유(주)  24060  2015/12                    15000000.0  3040.0   

               시가총액  
0      7.900370e+10  
1      4.611375e+11  
2      5.902920e+11  
3      8.027972e+11  
4      6.794564e+11  
...             ...  
17482  8.520000e+10  
17483  3.000000e+10  
17484  4.057500e+10  
17485  5.137500e+10  
17486  4.560000e+10  

[17487 rows x 6 columns]

df_price -> price 
- isin() 해서 하는 것보다 나은 방법 있는지 강구해볼 것 ! 

### 2. mk_real의 거래소코드에 해당하는 df_price 뽑기 -> price


In [14]:
# 2. price 데이터프레임
price = df_price.loc[df_price['거래소코드'].isin(mk_real['거래소코드'].astype('str').unique())]

In [15]:
# df_price -> price 넘어오면서, 우리가 원하는 기업 중 일부만 남는 듯.
## df_price에 애초에 기업이 전부 다 있는게 아니거나 위 code가 잘못되었거나 -> 확인 필요 
print(len(price['회사명'].unique()))
print(len(price['거래소코드'].unique()))

777
778


In [16]:
# price 시가총액 컬럼 추가
price['시가총액'] = price['종가(원)'] * price['상장주식수(주)']

C:\Users\zino0\AppData\Local\Temp\ipykernel_20652\1986805422.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  price['시가총액'] = price['종가(원)'] * price['상장주식수(주)']


In [17]:
# price 회계년도 -> 시계열
price['회계년도'] = pd.to_datetime(price['회계년도'], format='%Y/%m/%d')
price.info()

C:\Users\zino0\AppData\Local\Temp\ipykernel_20652\2639145216.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  price['회계년도'] = pd.to_datetime(price['회계년도'], format='%Y/%m/%d')


<class 'pandas.core.frame.DataFrame'>
Int64Index: 881097 entries, 1968 to 380385
Data columns (total 7 columns):
 #   Column    Non-Null Count   Dtype         
---  ------    --------------   -----         
 0   회사명       881097 non-null  object        
 1   거래소코드     881097 non-null  object        
 2   회계년도      881097 non-null  datetime64[ns]
 3   시가(원)     881097 non-null  int64         
 4   종가(원)     881097 non-null  int64         
 5   상장주식수(주)  881097 non-null  int64         
 6   시가총액      881097 non-null  int64         
dtypes: datetime64[ns](1), int64(4), object(2)
memory usage: 53.8+ MB


In [18]:
# 확인용 df_price회계년도 -> 시계열
df_price['회계년도'] = pd.to_datetime(df_price['회계년도'], format = '%Y/%m/%d')
# df_price.info()
print(df_price.loc[df_price['회계년도'] <= '2012-01-01']) # 1. 얘랑

print('\n')
print('------------------------------------------------------------')
print('\n')

price.loc[price['회계년도'] <= '2016-01-01']  # 2. 얘 비교 -> price에 2016 이후 data 없음 

                 회사명   거래소코드       회계년도  시가(원)  종가(원)  상장주식수(주)
0       3노드디지탈그룹유한공사  900010 2011-01-03   1870   1905  57664070
1       3노드디지탈그룹유한공사  900010 2011-01-04   1925   1970  57664070
2       3노드디지탈그룹유한공사  900010 2011-01-05   1980   1970  57664070
3       3노드디지탈그룹유한공사  900010 2011-01-06   1985   1995  57664070
4       3노드디지탈그룹유한공사  900010 2011-01-07   2010   1985  57664070
...              ...     ...        ...    ...    ...       ...
256436         히스토스템   36840 2011-09-06     59     61  29675245
256437         히스토스템   36840 2011-09-07     70     78  29675245
256438         히스토스템   36840 2011-09-08     83     77  29675245
256439         히스토스템   36840 2011-09-09    106     88  29675245
256440         히스토스템   36840 2011-09-14     65     55  29675245

[256441 rows x 6 columns]


------------------------------------------------------------




Empty DataFrame
Columns: [회사명, 거래소코드, 회계년도, 시가(원), 종가(원), 상장주식수(주), 시가총액]
Index: []

### *price에 2016 이후 data 없음 & 기업수도 777개 (종목코드는 778개)*
### *mk_real 은 각각 기업수 1818, 종목코드 1795*

---
### price 문제 있는 상태로 일단 진행

---

### 3. price에서 각 회계년도 마지막 주가 data만 남기기 


In [19]:
# 3. 회계년도 마지막 data

## 'Date' 컬럼에서 년도만 추출하여 새로운 'Year' 컬럼 추가
price['Year'] = price['회계년도'].dt.year


## 각 년도별로 그룹화하고 마지막 데이터를 선택
# last_data_by_year = price.groupby(['Year']).last().reset_index(drop=True)
# last_data_by_year = price.groupby(['회사명','거래소코드','회계년도','Year']).last().reset_index()

last_data_by_year = price.groupby(['회사명', '거래소코드', 'Year']).last().reset_index()

last_data_by_year.drop(['Year', '시가(원)'], axis =1, inplace = True )

price_last = last_data_by_year.copy()


C:\Users\zino0\AppData\Local\Temp\ipykernel_20652\2586334058.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  price['Year'] = price['회계년도'].dt.year


In [20]:
price_last

회사명   거래소코드       회계년도  종가(원)  상장주식수(주)          시가총액
0        가온칩스  399720 2022-12-29  14200  11488320  163134144000
1     강스템바이오텍  217730 2016-12-29  10800  14213650  153507420000
2     강스템바이오텍  217730 2017-12-28  12500  16251712  203146400000
3     강스템바이오텍  217730 2018-12-28  21800  20824937  453983626600
4     강스템바이오텍  217730 2019-12-30   8140  20849437  169714417180
...       ...     ...        ...    ...       ...           ...
3885       힘스  238490 2018-12-28   9480   5656118   53619998640
3886       힘스  238490 2019-12-30  27200   5656118  153846409600
3887       힘스  238490 2020-12-30  12800  11312236  144796620800
3888       힘스  238490 2021-12-30   9850  11312236  111425524600
3889       힘스  238490 2022-12-29   6060  11312236   68552150160

[3890 rows x 6 columns]

### 4. price_last 로 mk_real 시총 Update
    - price_last 의 시총 컬럼으로 mk_real 에 합쳐서(merge) 결측치 줄어드는지 확인 

In [21]:
# merge 전 data type 일치 
price_last['거래소코드'] = price_last['거래소코드'].astype('int')
mk_real['회계년도'] = pd.to_datetime(mk_real['회계년도'], format = '%Y/%m/%d')


# price_last.info()
# mk_real.info()

C:\Users\zino0\AppData\Local\Temp\ipykernel_20652\2755046273.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  mk_real['회계년도'] = pd.to_datetime(mk_real['회계년도'], format = '%Y/%m/%d')


In [22]:
# merge 전 컬럼이름 구분 위해 컬럼이름 변경
price_last.rename(columns = {'시가총액':'시가총액_x'}, inplace = True)

In [23]:
# A와 B 데이터프레임을 '회사명', '거래소코드', '회계년도' 컬럼을 기준으로 병합 (left join)
mkcap = pd.merge(mk_real, price_last, on=['회사명', '거래소코드', '회계년도'], how='left')

# 'x' 컬럼의 결측치를 'y' 컬럼 값으로 채움
mkcap['시가총액'] = mkcap['시가총액'].fillna(mkcap['시가총액_x'])

# 불필요한 'y' 컬럼 삭제
mkcap.drop(columns='시가총액_x', inplace=True)

# 불필요한 다른 컬럼 삭제
mkcap.drop(['종가(원)_x','종가(원)_y', '상장주식수(주)'], axis = 1, inplace = True)

In [24]:
mkcap.isna().sum()

회사명                                0
거래소코드                              0
회계년도                               0
[제조]* 발행한 주식총수(*)(IFRS연결)(주)    1875
시가총액                            3478
dtype: int64

In [29]:
mkcap.rename(columns={'[제조]* 발행한 주식총수(*)(IFRS연결)(주)':
              '발행주식수(연결+개별)'}, inplace=True)

In [31]:
mkcap.to_excel('../dataset_2nd/mkcap_2.xlsx')

---

1. 시총뽑기 
    - df_code의 0:4 까지 컬러만 따로 df : mkcap (= Market Capitalization, 시가총액)
    - df_code의 거래소코드에 해당하는 df_price 뽑기 -> price
    - price에서 각 회계년도 마지막 주가 data만 남기기 
    - price에서 종가 x 상장주식수( = 보통주 + 우선주) 로 시총 df 따로 만들기 

2. 시총 to Yuna

3. Z-score Update
    - New 시총 사용해서 (결측치 확인)
    - 저량변수 평균값 사용